File: Knapsack01.py
import numpy as np
def KnapSack(values, weights,capacity):
    n = len(values)
    dm = np.zeros((n + 1, capacity + 1))
    for i in range(0, n):
        for j in range(1, capacity + 1):
            if weights[i] <= j:
                cost = values[i] + dm[i][j - weights[i]]
                dm[i+1][j] = max(dm[i][j], cost)
            else:
                dm[i+1][j] = dm[i][j]
    
    selected_items = []
    i, j = n, capacity 
    while i > 0 and j > 0:
        if dm[i][j] != dm[i - 1][j]:
            selected_items.append(i - 1)
            j-=weights[i-1]
        i-=1
    return selected_items, dm

if __name__ == "__main__":
    keys = ["A", "B", "C", "D", "E", "F"]
    values = [1,2,5,6,9,4]
    weights = [2,3,4,5,10,3]
    selected_items, data_matrix = KnapSack(values, weights, 15)

    selected_keys = []
    for i in selected_items:
        selected_keys.append(keys[i])

    print(data_matrix)
    print()
    print(selected_keys)
 

File: KnapSack01.cpp
#include<iostream>
#include<vector>
using namespace std;
vector<vector<int>> KnapSack( int* values, int* weights, int n, int capacity)
{
    vector<vector<int>> A(n+1, vector<int>(capacity+1, 0));   

    for(int i = 0; i < n; i++)
    {
        for(int j = 1; j < capacity + 1; j++)
        {
            if (weights[i] <= j)
            {
                int cost = values[i] + A[i][j - weights[i]];
                A[i+1][j] = max(A[i][j], cost);
            }
            else
            {
                A[i+1][j] = A[i][j];
            }
        }
    }
    return A;
}